presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -111,10 +132,39 @@ public void process(Iterable<QueryExecution> queries)
             totalBytes += bytes;
             if (bytes > queryMemoryLimit) {
                 query.fail(new ExceededMemoryLimitException("Query", DataSize.succinctDataSize(queryMemoryLimit, Unit.BYTE)));
+                queryKilled = true;
             }
         }
         clusterMemoryUsageBytes.set(totalBytes);
 
+        boolean shouldKillQuery = nanosSince(lastTimeNotOutOfMemory).compareTo(killOnOutOfMemoryDelay) == 1 && isClusterOutOfMemory();
+        boolean lastKilledQueryIsGone = lastKilledQuery == null;
+
+        if (lastKilledQuery != null) {
+            ClusterMemoryPool generalPool = pools.get(GENERAL_POOL);
+            if (generalPool != null) {
+                lastKilledQueryIsGone = generalPool.getQueryMemoryReservations().containsKey(lastKilledQuery);
+            }
+        }
+
+        if (shouldKillQuery && lastKilledQueryIsGone && !queryKilled && killOnOutOfMemory) {
+            // Kill the biggest query in the general pool
+            QueryExecution biggestQuery = null;
+            long maxMemory = -1;
+            for (QueryExecution query : queries) {
+                long bytesUsed = query.getTotalMemoryReservation();
+                if (bytesUsed > maxMemory && query.getMemoryPool().getId().equals(GENERAL_POOL)) {
+                    biggestQuery = query;
+                    maxMemory = bytesUsed;
+                }
+            }
+            if (biggestQuery != null) {
+                biggestQuery.fail(new PrestoException(CLUSTER_OUT_OF_MEMORY, "The cluster is out of memory, and your query was killed. Please try again in a few minutes."));
```

Comment:
@cberner  How about to replace "your query" with the specific query information(id, context, etc)?  It might be more meaningful (helpful) for the people who submitted the query.  
Another question is:  if this "biggestQuery" has higher priority (e.g. we want to sacrifice other smaller queries to get the result of "biggestQuery" first),  is there any config or option? Thanks.